### PR TITLE
Chore!: Introduce the snapshot cache to the state sync

### PR DESCRIPTION
**Pull request by izeigerman**
This update introduces a file cache for snapshots fetched from the state, which should significantly speed up plan generation time.

### CODE REVIEW

File: sqlmesh/core/snapshot/cache.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+import typing as t
+
+from pathlib import Path
+from sqlmesh.core.model.cache import OptimizedQueryCache
+from sqlmesh.core.snapshot.definition import Snapshot, SnapshotId
+from sqlmesh.utils.cache import FileCache
+
+
+class SnapshotCache:
```
Comment (tobymao) [inline sqlmesh/core/snapshot/cache.py:11]: should this not be in the state sync cache?

File: sqlmesh/core/state_sync/engine_adapter.py [hunk 2]
```diff
@@ -212,6 +213,8 @@ def push_snapshots(self, snapshots: t.Iterable[Snapshot]) -> None:
 
         if snapshots:
             self._push_snapshots(snapshots)
+            for snapshot in snapshots:
+                self._snapshot_cache.put(snapshot)
```
Comment (tobymao) [inline sqlmesh/core/state_sync/engine_adapter.py:217]: is this going to reload the optimized query cache even if it's not necessary? we should make sure with_optimized_query only loads when necessary

File: sqlmesh/core/snapshot/cache.py (hunk 1 above)
Comment (izeigerman) [reply sqlmesh/core/snapshot/cache.py:11]: I think this is pretty independent from the context in which it's used.

File: sqlmesh/core/state_sync/engine_adapter.py (hunk 2 above)
Comment (izeigerman) [reply sqlmesh/core/state_sync/engine_adapter.py:217]: Wdym by "reload"? Reading from the file system?

**Review by tobymao — APPROVED** [review]

**Pull request merged**